### ISSUE TITLE
Struggling to understand ChanceMode

### ISSUE THREAD

**Issue by aaronsnoswell** [deepmind/open_spiel#792]
Hi OpenSpiel team!

I'm excited about using OpenSpiel as a tool when prototyping my own board game designs.

> TLDR; when should I use `ChanceMode::kExplicitStochastic` or `ChanceMode::kSampledStochastic` for for a stochastic game?

I'm attempting to define my first (non-deterministic) game, and I'm struggling to understand the role of the `ChanceMode` enum.

`spiel.h:67` [has the following](https://github.com/deepmind/open_spiel/blob/5eaf401f7d08d68285fe214ab5cf5a741807ea6e/open_spiel/spiel.h#L67);

``` 
  // Are there any chance nodes? If so, how is chance treated?
  // Either all possible chance outcomes are explicitly returned as
  // ChanceOutcomes(), and the result of ApplyAction() is deterministic. Or
  // just one ChanceOutcome is returned, and the result of ApplyAction() is
  // stochastic.
  enum class ChanceMode {
    kDeterministic,       // No chance nodes
    kExplicitStochastic,  // Has at least one chance node, all with
                          // deterministic ApplyAction()
    kSampledStochastic,   // At least one chance node with non-deterministic
                          // ApplyAction()
  };
  ChanceMode chance_mode;
```

And I can also see some information about chance nodes in the docstring for `ApplyAction()` at [`spiel.h:225`](https://github.com/deepmind/open_spiel/blob/5eaf401f7d08d68285fe214ab5cf5a741807ea6e/open_spiel/spiel.h#L225);

```
  // Change the state of the game by applying the specified action in turn-based
  // games or in non-simultaneous nodes of simultaneous move games.
  // This function encodes the logic of the game rules.
  //
  // In the case of chance nodes, the behavior of this function depends on
  // GameType::chance_mode. If kExplicit, then the outcome should be
  // directly applied. If kSampled, then a dummy outcome is passed and the
  // sampling of and outcome should be done in this function and then applied.
  //
  // Games should implement DoApplyAction.
  virtual void ApplyAction(Action action_id);
```

---

From these descriptions, it seems like you can use either `kExplicit` or `kSampeld` - it's just a matter of how your implementation code works. Is my understanding correct? Are there stochastic games that can only be implemented with one or the other?

I can see that very few games used the sampled mode;

```
import pyspiel

# All games with kSampledStochastic chance mode.
SPIEL_SAMPLED_STOCHASTIC_GAMES_LIST = [
    g for g in pyspiel.registered_games() if g.default_loadable and
    g.chance_mode == pyspiel.GameType.ChanceMode.SAMPLED_STOCHASTIC
]

SPIEL_SAMPLED_STOCHASTIC_GAMES_LIST
```

```
[<GameType 'bridge_uncontested_bidding'>,
 <GameType 'negotiation'>,
 <GameType 'stones_and_gems'>,
 <GameType 'tarok'>]
```

But don't know enough about these games to tell why they might chose the sampled mode rather than the explicit mode.

I can't find anything about `ChanceMode` in the OpenSpiel paper, and I couldn't find any other documentation to answer this question - perhaps a documentation section could be added for this.

**Comment by lanctot** [comment]
Hi @aaronsnoswell, 

Good question: it is subtle. After you understand it I would appreciate any advice you have on improving the docs and encourage you to do that!

Firstly: most of the games use `kExplicitStochastic`, it allows greater flexibility, and several algorithms need the information. When your game is explicit stochastic, it means that the implementation has to provide the distribution of chance event outcomes and the logic on how each is applied. Example: Backgammon. When a player is done their turn, the state is at a chance nodes. That chance node has a number of outcomes (corresponding to the different ways both dice can be rolled), and each has a probability. As the person implementing the game, you need to provide both the set of outcomes and probability of each (`State::ChanceOutcomes`) and the logic in `ApplyAction` that knows how to update the internal state for each outcome. 

You can use `kSampledStochastic` when computing or representing this distribution explicitly or encoding the rules logic it into the rules is awkward or inefficient, and normally in a domain that you don't plan to use search or other dynamic programming / value iteration or tabular methods for (i.e. just sample-based RL). In this case, at chance nodes, you have only one legal outcome (without any probability associated to it), and when you apply that action, the state is changed based on a sampling from a random number generator stored in the game. So you never get access to the probabilities, and you can only sample chance outcomes (not enumerate them).

It is better to implement the game using explicit stochastic if you can, because then more algorithms can be applied to the game. But if you only plan to use the game for RL, probably ok to do sampled stochastic. What game are you thinking about? Because the best thing would be to find the game that most closely matches yours and check how it's done in that game. But also happy to answer more questions about it if it's still unclear.

**Comment by aaronsnoswell** [comment]
Aaaah I understand now. Thanks @lanctot!

So explicit mode communicates the full support of the chance outcome distribution to a learning algorithm, while sampled mode just passed the sampled outcome.

I will create a PR for some extra documentation explaining this. Would it be ok to create a new `.md` file in the `/docs` folder if necessary to do this?

**Comment by lanctot** [comment]
> So explicit mode communicates the full support of the chance outcome distribution to a learning algorithm, while sampled mode just passes the sampled outcome.

Yes, exactly. But to clarify: the second still uses chance nodes, it's just that every chance node as having a single legal outcome. And it's only when you try to call `ApplyAction` using that single outcome that the state changes (stochastically, using a random number generator).
 
> I will create a PR for some extra documentation explaining this. Would it be ok to create a new `.md` file in the `/docs` folder if necessary to do this?

Absolutely, thanks! I would love for it to be linked from somewhere, though. Possible in the comments of the `spiel.h` where chance mode is described or in the developer guide (or both).

**Comment by aaronsnoswell** [comment]
Great. Thanks for the clarification.

> Absolutely, thanks! I would love for it to be linked from somewhere, though. Possible in the comments of the spiel.h where chance mode is described or in the developer guide (or both).

I'll give it a crack.

As for games - I'm presently implementing [Battle Line](https://boardgamegeek.com/boardgame/760/battle-line) as a learning exercise (it is similar to a game design I'm working on). It's a two-player zero-sum sequential partial information card game. I'm using python for now for quick prototyping.

---

This discussion has reminded me of a related follow-up question; say I use the Explicit Stochastic ChanceMode, does the chance node vector of outcome-probability tuples that `chance_outcomes()` returns have to be of a fixed size for all chance nodes? Presently, my implementation returns a different sized vector of outcome-probability tuples based on the current game state, but I wonder if this vector of outcome-probability tuples should actually be of a fixed size for all possible chance nodes, and I just set the probabilities to zero for the chance outcomes that are illegal/impossible given the current state.

Can you clarify the intended convention here?

Thanks!

**Comment by lanctot** [comment]
Battle Line is a great choice of game to start with! (I'm quite a fan of Reiner Knizia.)

> This discussion has reminded me of a related follow-up question; say I use the Explicit Stochastic ChanceMode, does the chance node vector of outcome-probability tuples that `chance_outcomes()` returns have to be of a fixed size for all chance nodes?

Nope, it's perfectly fine (and expected) to have different number of outcomes at each chance node. For example, the poker games have a sequence of chance nodes where each one deals one card and then the deck reduces, so each successor chance node has one has 1 less outcome than its predecessor.
